lib/proxy/client.go

```diff
@@ -0,0 +1,563 @@
+// Copyright 2022 Gravitational, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"sync"
+	"time"
+
+	clientapi "github.com/gravitational/teleport/api/client/proto"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/metadata"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+)
+
+// ClientConfig configures a Client instance.
+type ClientConfig struct {
+	// Context is a signalling context
+	Context context.Context
+	// ID is the ID of this server proxy
+	ID string
+	// AuthClient is an auth client
+	AuthClient auth.ClientI
+	// AccessPoint is a caching auth client
+	AccessPoint auth.ProxyAccessPoint
+	// TLSConfig is the proxy client TLS configuration.
+	TLSConfig *tls.Config
+	// Log is the proxy client logger.
+	Log logrus.FieldLogger
+	// Clock is used to control connection monitoring ticker.
+	Clock clockwork.Clock
+	// GracefulShutdownTimout is used set the graceful shutdown
+	// duration limit.
+	GracefulShutdownTimeout time.Duration
+
+	// getConfigForServer updates the client tls config.
+	// configurable for testing purposes.
+	getConfigForServer func() (*tls.Config, error)
+
+	// sync runs proxy and connection syncing operations
+	// configurable for testing purposes
+	sync func()
+}
+
+// checkAndSetDefaults checks and sets default values
+func (c *ClientConfig) checkAndSetDefaults() error {
```

Comment:
Is there a reason for making this private? 99% of these are public in the codebase and it seems a bit off to me that `ClientConfig` is public yet I wouldn't be able to verify it was populated properly by calling this.

Reply:
There's no reason at all, I must have written it without checking how it's done elsewhere.
But it is my understanding that calling `CheckAndSetDefaults` is implied in every `NewSomething` call. `ClientConfig` is worthless on its own without the `NewClient` call, so I wouldn't expect anyone to make the check manually. Unless services start sharing the same config structs or the function is not automatically called when initializing a new object, I don't see why it should be public. Am I missing something?